**<h1 align="center" > Business Cases with Data Science - Case 1</h1>**


**Group members:**<br> 
Chloé Deschanel - 20240693 <br>
Diogo Carvalho - 20240694 <br>
Ingrid Lopez - 20240692 <br>
Ruben Marques- 20240352

# Table of Contents

# 1. Data Exploration

 ## 1.1. Imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## 1.2 Loading Data

In [2]:
data = pd.read_csv("../Data/Case1_HotelCustomerSegmentation.csv", sep=";")

In [3]:
# Display all columns
pd.set_option('display.max_columns', None)
data.sample(5)

ID Nationality   Age  DaysSinceCreation  \
37787    37788         CHE  31.0                746   
9044      9045         SRB  48.0               1208   
100049  100050         EGY  50.0                131   
4726      4727         FRA  53.0               1278   
18079    18080         DNK  37.0               1071   

                                                 NameHash  \
37787   0xE7A5AF0993B3A7791A91FE801D005653CF80F42AF66D...   
9044    0x3F31D322B341CECC1877BCE9D794DE3D730DD55FB194...   
100049  0x3ED38CEC5DBEA14966747C4DBEC500606603F5FC17A5...   
4726    0x8082EABD1951CFF5DE0F8AA1C115A4BAB16B0C9456CB...   
18079   0x627B613C860B1BE37F19C263EAA13D4F9AEB48E86E19...   

                                                DocIDHash  AverageLeadTime  \
37787   0xD3BC077CC256136A5B2A47F307F5E1B193140BC9DE3A...              263   
9044    0x71A016592DE2AEA5600498E6CA6359B314077BBB8AE7...              192   
100049  0xEBBA92B08C7EF5A6F160115C338EDF7D0621A1447B3A...               20   
4726    0x601FF8BE46060443B729652315C4DFD1D8F4EB1041B1...               26   
18079   0x51E4723F010E31C92881935D6D047DF9EDD771CC6744...               81   

        LodgingRevenue  OtherRevenue  BookingsCanceled  BookingsNoShowed  \
37787            311.4          79.5                 0                 0   
9044             178.0          18.0                 0                 0   
100049           668.8          32.0                 0                 0   
4726             780.0         105.0                 0                 0   
18079            562.4          42.0                 0                 0   

        BookingsCheckedIn  PersonsNights  RoomNights    DistributionChannel  \
37787                   1              9           3  Travel Agent/Operator   
9044                    1              2           2  Travel Agent/Operator   
100049                  1              4           4  Travel Agent/Operator   
4726                    1              6           2  Travel Agent/Operator   
18079                   1              4           4  Travel Agent/Operator   

       MarketSegment  SRHighFloor  SRLowFloor  SRAccessibleRoom  \
37787          Other            0           0                 0   
9044          Groups            0           0                 0   
100049         Other            0           0                 0   
4726           Other            0           0                 0   
18079          Other            0           0                 0   

        SRMediumFloor  SRBathtub  SRShower  SRCrib  SRKingSizeBed  SRTwinBed  \
37787               0          0         0       0              1          0   
9044                0          0         0       0              0          0   
100049              0          0         0       0              1          0   
4726                0          0         0       0              1          0   
18079               0          0         0       0              1          0   

        SRNearElevator  SRAwayFromElevator  SRNoAlcoholInMiniBar  SRQuietRoom  
37787                0                   0                     0            0  
9044                 0                   0                     0            0  
100049               0                   0                     0            0  
4726                 0                   0                     0            0  
18079                0                   0                     0            0

In [4]:
# Set ID as index
data.set_index('ID', inplace=True)

| #   | Column Name        | Description                                                                                       |
|:----|:-------------------|:--------------------------------------------------------------------------------------------------|
| 1   | ID        | Customer ID customer. |
| 2   | Nationality    | Nationality of the customer in ISO 3166-1 (Alpha 3) format. |
| 3   | Age       | Age of the customer. |
| 4   | DaysSinceCreation       | Number of elapsed days since the customer was created. |
| 5   | NameHash | Hash of the customer's name. |
| 6   | DocIDHash | Hash of the customer’s personal document identification number (usually a passport or ID card). |
| 7   | AverageLeadTime | Average number of days before arrival date the customer makes bookings. |
| 8   | LodgingRevenue | Total amount of lodging revenue paid by the customer so far. |
| 9   | OtherRevenue | Total amount of other revenue (e.g., food & beverage, spa, etc.) paid by the customer so far. |
| 10  | BookingsCanceled | Number of bookings the customer made but subsequently canceled. |
| 11  | BookingsNoShowed | Number of bookings the customer made but subsequently made a "no-show". |
| 12  | BookingsCheckedin | Number of bookings the customer made, which actually ended up staying. |
| 13  | PersonNights | Total person/nights the customer has stayed at the hotel so far. Persons/Nights are the sum of Adults and Children in each booking, multiplied by the number of Nights (Length-of-stay) of the booking. |
| 14  | RoomNights | Total of room/nights the customer has stayed at the hotel so far. Room/Nights are the multiplication of the number of rooms of each booking by the the number of Nights (Length-of- stay) of the booking. |
| 15  | DistributionChannel | Distribution channel normally used by the customer to make bookings at the hotel. |
| 16  | MarketSegment | Current market segment of the customer. |
| 17  | SRHighFloor | Indication if the customer usually asks for a room in a higher floor (0: No, 1: Yes). |
| 18  | RLowFloor  | Indication if the customer usually asks for a room in a lower floor (0: No, 1: Yes).  | 
| 19  | SRAccessibleRoom | Indication if the customer usually asks for an accessible room (0: No, 1: Yes). |
| 20  | SRMediumFloor | Indication if the customer usually asks for a room in a middle floor (0: No, 1: Yes). |
| 21  | SRBathtub | Indication if the customer usually asks for a room with a bathtub (0: No, 1: Yes). |
| 22  | SRShower | Indication if the customer usually asks for a room with a shower (0: No, 1: Yes). |
| 23  | SRCrib  | Indication if the customer usually asks for a crib (0: No, 1: Yes). |
| 24  | SRKingSizeBed | Indication if the customer usually asks for a room with a king size bed (0: No, 1: Yes). |
| 25  | SRTwinBed | Indication if the customer usually asks for a room with a twin bed (0: No, 1: Yes). |
| 26  | SRNearElevator | Indication if the customer usually asks for a room near the elevator (0: No, 1: Yes). |
| 27  | SRAwayFromElevator | Indication if the customer usually asks for a room away from the elevator (0: No, 1: Yes). |
| 28  | SRNoAlcoholInMiniBar | Indication if the customer usually asks for a room with no alcohol in the mini bar (0: No, 1: Yes). |
| 29  | SRQuietRoom | Indication if the customer usually asks for a room away from the noise (0: No, 1: Yes). |


# Data analysis

In [5]:
# Drop columns relative to identification (NameHash, DocIDHash)
data.drop(['NameHash', 'DocIDHash'], axis=1, inplace=True)

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 111733 entries, 1 to 111733
Data columns (total 26 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Nationality           111733 non-null  object 
 1   Age                   107561 non-null  float64
 2   DaysSinceCreation     111733 non-null  int64  
 3   AverageLeadTime       111733 non-null  int64  
 4   LodgingRevenue        111733 non-null  float64
 5   OtherRevenue          111733 non-null  float64
 6   BookingsCanceled      111733 non-null  int64  
 7   BookingsNoShowed      111733 non-null  int64  
 8   BookingsCheckedIn     111733 non-null  int64  
 9   PersonsNights         111733 non-null  int64  
 10  RoomNights            111733 non-null  int64  
 11  DistributionChannel   111733 non-null  object 
 12  MarketSegment         111733 non-null  object 
 13  SRHighFloor           111733 non-null  int64  
 14  SRLowFloor            111733 non-null  int64  
 15  SRAcc

In [7]:
data.describe().T

count        mean         std   min    25%    50%  \
Age                   107561.0   45.639191   17.244952 -10.0   33.0   47.0   
DaysSinceCreation     111733.0  595.026599  374.657382  36.0  288.0  522.0   
AverageLeadTime       111733.0   60.833147   85.115320  -1.0    0.0   21.0   
LodgingRevenue        111733.0  283.851283  379.131556   0.0    0.0  208.0   
OtherRevenue          111733.0   64.682802  123.580715   0.0    0.0   31.0   
BookingsCanceled      111733.0    0.002282    0.080631   0.0    0.0    0.0   
BookingsNoShowed      111733.0    0.000600    0.028217   0.0    0.0    0.0   
BookingsCheckedIn     111733.0    0.737607    0.730889   0.0    0.0    1.0   
PersonsNights         111733.0    4.328318    4.630739   0.0    0.0    4.0   
RoomNights            111733.0    2.203825    2.301637   0.0    0.0    2.0   
SRHighFloor           111733.0    0.042512    0.201755   0.0    0.0    0.0   
SRLowFloor            111733.0    0.001307    0.036125   0.0    0.0    0.0   
SRAccessibleRoom      111733.0    0.000224    0.014957   0.0    0.0    0.0   
SRMediumFloor         111733.0    0.000770    0.027733   0.0    0.0    0.0   
SRBathtub             111733.0    0.003132    0.055881   0.0    0.0    0.0   
SRShower              111733.0    0.001629    0.040327   0.0    0.0    0.0   
SRCrib                111733.0    0.016181    0.126173   0.0    0.0    0.0   
SRKingSizeBed         111733.0    0.363268    0.480943   0.0    0.0    0.0   
SRTwinBed             111733.0    0.156811    0.363624   0.0    0.0    0.0   
SRNearElevator        111733.0    0.000331    0.018195   0.0    0.0    0.0   
SRAwayFromElevator    111733.0    0.003598    0.059874   0.0    0.0    0.0   
SRNoAlcoholInMiniBar  111733.0    0.000197    0.014031   0.0    0.0    0.0   
SRQuietRoom           111733.0    0.087718    0.282886   0.0    0.0    0.0   

                        75%       max  
Age                    58.0    123.00  
DaysSinceCreation     889.0   1385.00  
AverageLeadTime        95.0    588.00  
LodgingRevenue        393.3  21781.00  
OtherRevenue           84.0   8859.25  
BookingsCanceled        0.0     15.00  
BookingsNoShowed        0.0      3.00  
BookingsCheckedIn       1.0     76.00  
PersonsNights           6.0    116.00  
RoomNights              3.0    185.00  
SRHighFloor             0.0      1.00  
SRLowFloor              0.0      1.00  
SRAccessibleRoom        0.0      1.00  
SRMediumFloor           0.0      1.00  
SRBathtub               0.0      1.00  
SRShower                0.0      1.00  
SRCrib                  0.0      1.00  
SRKingSizeBed           1.0      1.00  
SRTwinBed               0.0      1.00  
SRNearElevator          0.0      1.00  
SRAwayFromElevator      0.0      1.00  
SRNoAlcoholInMiniBar    0.0      1.00  
SRQuietRoom             0.0      1.00

## Age
`Age:` <br>
Has missing values; <br>
Average age is 45 years; <br>
Min is -10; <br>
Max is 123.

Actionables: Deal with missing values, plot age distribution, check cases where age is negative, check cases where age is above 100.

In [8]:
# Display rows for missing values in age column
data[data['Age'].isnull()]

Nationality  Age  DaysSinceCreation  AverageLeadTime  LodgingRevenue  \
ID                                                                            
2              PRT  NaN               1385               61          280.00   
21             PRT  NaN               1385               63          224.00   
39             PRT  NaN               1385                0            0.00   
40             PRT  NaN               1385                0            0.00   
56             PRT  NaN               1385               97          213.52   
...            ...  ...                ...              ...             ...   
111549         PRT  NaN                 39                0            0.00   
111553         PRT  NaN                 38              204          381.60   
111631         PRT  NaN                 37                0            0.00   
111634         PRT  NaN                 37                0            0.00   
111635         PRT  NaN                 37                0            0.00   

        OtherRevenue  BookingsCanceled  BookingsNoShowed  BookingsCheckedIn  \
ID                                                                            
2               53.0                 0                 0                  1   
21             156.0                 0                 0                  1   
39               0.0                 0                 0                  1   
40               0.0                 0                 0                  1   
56              11.5                 0                 0                  1   
...              ...               ...               ...                ...   
111549           0.0                 0                 0                  0   
111553          91.2                 0                 0                  1   
111631           0.0                 0                 0                  0   
111634           0.0                 0                 0                  0   
111635           0.0                 0                 0                  0   

        PersonsNights  RoomNights    DistributionChannel  \
ID                                                         
2                  10           5  Travel Agent/Operator   
21                  4           2  Travel Agent/Operator   
39                  1           1                 Direct   
40                  1           1                 Direct   
56                  8           4  Travel Agent/Operator   
...               ...         ...                    ...   
111549              0           0  Travel Agent/Operator   
111553              8           4  Travel Agent/Operator   
111631              0           0  Travel Agent/Operator   
111634              0           0  Travel Agent/Operator   
111635              0           0  Travel Agent/Operator   

                MarketSegment  SRHighFloor  SRLowFloor  SRAccessibleRoom  \
ID                                                                         
2       Travel Agent/Operator            0           0                 0   
21                      Other            0           0                 0   
39              Complementary            0           0                 0   
40              Complementary            0           0                 0   
56      Travel Agent/Operator            0           0                 0   
...                       ...          ...         ...               ...   
111549                  Other            0           0                 0   
111553  Travel Agent/Operator            0           0                 0   
111631                  Other            0           0                 0   
111634                  Other            0           0                 0   
111635                  Other            0           0                 0   

        SRMediumFloor  SRBathtub  SRShower  SRCrib  SRKingSizeBed  SRTwinBed  \
ID                                                                             
2                   0

In [9]:
# Check age correlation with other columns 
data.select_dtypes(include=[np.number]).corr()['Age'].sort_values(ascending=False)

Age                     1.000000
AverageLeadTime         0.261786
BookingsCheckedIn       0.146982
RoomNights              0.139599
DaysSinceCreation       0.127568
PersonsNights           0.120793
OtherRevenue            0.118851
LodgingRevenue          0.061177
SRTwinBed               0.050583
SRNearElevator          0.013616
SRLowFloor              0.012960
SRShower                0.008962
BookingsCanceled        0.008759
SRQuietRoom             0.008154
SRAwayFromElevator      0.007683
BookingsNoShowed        0.007069
SRAccessibleRoom        0.004509
SRMediumFloor           0.003319
SRNoAlcoholInMiniBar   -0.011765
SRHighFloor            -0.021887
SRBathtub              -0.022021
SRKingSizeBed          -0.025645
SRCrib                 -0.136364
Name: Age, dtype: float64

Our goal by checking this correlation was to see if any variable was correlated with age enough, that we could use in a predictive model to predict age. Since the highest correlated variable is 26% correlation, we discarded this idea, and just assumed to input median.

In [10]:
# Check cases where age is negative
print(f'There are {len(data[data['Age'] < 0])} cases where age is negative')
data[data['Age'] < 0]

There are 14 cases where age is negative


Nationality   Age  DaysSinceCreation  AverageLeadTime  LodgingRevenue  \
ID                                                                            
2053          BIH  -6.0               1324                0            0.00   
2469          ATF  -6.0               1319                0            0.00   
5128          FRA  -8.0               1271               66           93.30   
6752          GBR -10.0               1243               22          450.00   
8361          DEU  -6.0               1217              256          126.00   
8629          NLD  -9.0               1213              111          414.60   
9920          DZA  -5.0               1193               54          138.93   
10663         DEU  -9.0               1181              292          126.00   
10695         DEU  -9.0               1181              292          128.00   
10756         PRT  -5.0               1181              292          128.00   
14322         PRT  -9.0               1127              346          128.00   
15105         DEU -10.0               1117              276          240.00   
20407         USA  -8.0               1022              163          939.60   
57358         PRT  -5.0                510              130          388.80   

       OtherRevenue  BookingsCanceled  BookingsNoShowed  BookingsCheckedIn  \
ID                                                                           
2053           0.00                 0                 0                  0   
2469           0.00                 0                 0                  0   
5128          14.00                 0                 0                  1   
6752          18.00                 0                 0                  1   
8361         109.00                 0                 0                  1   
8629          34.00                 0                 0                  1   
9920          29.50                 0                 0                  1   
10663        113.00                 0                 0                  1   
10695         57.50                 0                 0                  1   
10756         55.00                 0                 0                  1   
14322         53.00                 0                 0                  1   
15105         61.25                 0                 0                  1   
20407        303.30                 0                 0                  1   
57358         47.20                 0                 0                  1   

       PersonsNights  RoomNights    DistributionChannel  \
ID                                                        
2053               0           0  Travel Agent/Operator   
2469               0           0  Travel Agent/Operator   
5128               2           1  Travel Agent/Operator   
6752               1           1  Travel Agent/Operator   
8361               4           2  Travel Agent/Operator   
8629               4           4  Travel Agent/Operator   
9920               4           1  Travel Agent/Operator   
10663              4           2  Travel Agent/Operator   
10695              2           2  Travel Agent/Operator   
10756              2           2  Travel Agent/Operator   
14322              2           2  Travel Agent/Operator   
15105              3           3  Travel Agent/Operator   
20407              9           3  Travel Agent/Operator   
57358              8           4  Travel Agent/Operator   

               MarketSegment  SRHighFloor  SRLowFloor  SRAccessibleRoom  \
ID                                                                        
2053   Travel Agent/Operator            0           0                 0   
2469   Travel Agent/Operator            0           0                 0   
5128                   Other            0           0                 0   
6752                   Other            0           0                 0   
8361                  Groups            0           0                 0   
8629                  

All rows with negative values for age have in common DistributionChannel = 'Travel Agent/Operator', and are between [-10, -5].


In [11]:
# Check cases where age is above 100
print(f'There are {len(data[data['Age'] > 100])} cases where age is above 100')
data[data['Age'] > 100]

There are 9 cases where age is above 100


Nationality    Age  DaysSinceCreation  AverageLeadTime  LodgingRevenue  \
ID                                                                             
3282          PRT  115.0               1305               22            79.0   
18912         BEL  111.0               1058              215           140.0   
25590         DEU  114.0                924                7           492.0   
44525         DEU  114.0                627                7           234.0   
45089         PRT  115.0                620                2           266.0   
53298         ITA  123.0                543                0             0.0   
58052         AUT  114.0                503                0             0.0   
64256         ESP  110.0                456                0             0.0   
87089         PRT  111.0                244               18           210.0   

       OtherRevenue  BookingsCanceled  BookingsNoShowed  BookingsCheckedIn  \
ID                                                                           
3282           25.5                 0                 0                  1   
18912         164.8                 0                 0                  1   
25590          76.4                 0                 0                  1   
44525          34.5                 0                 0                  1   
45089          43.0                 0                 0                  1   
53298           0.0                 0                 0                  0   
58052           0.0                 0                 0                  0   
64256           0.0                 0                 0                  0   
87089          52.0                 0                 0                  1   

       PersonsNights  RoomNights    DistributionChannel MarketSegment  \
ID                                                                      
3282               2           1  Travel Agent/Operator         Other   
18912              2           2              Corporate        Groups   
25590              4           4                 Direct        Direct   
44525              4           2  Travel Agent/Operator         Other   
45089              2           2  Travel Agent/Operator         Other   
53298              0           0  Travel Agent/Operator         Other   
58052              0           0  Travel Agent/Operator         Other   
64256              0           0  Travel Agent/Operator         Other   
87089              4           2                 Direct        Direct   

       SRHighFloor  SRLowFloor  SRAccessibleRoom  SRMediumFloor  SRBathtub  \
ID                                                                           
3282             0           0                 0              0          0   
18912            0           0                 0              0          0   
25590            0           0                 0              0          0   
44525            0           0                 0              0          0   
45089            1           0                 0              0          0   
53298            0           0                 0              0          0   
58052            0           0                 0              0          0   
64256            0           0                 0              0          0   
87089            1           0                 0              0          0   

       SRShower  SRCrib  SRKingSizeBed  SRTwinBed  SRNearElevator  \
ID                                                                  
3282          0       0              1          0               0   
18912         0       0              0          0               0   
25590         0       0              0          0               0   
44525         0       0              1          0               0   
45089         0       0              1          0               0   
53298         0       0              0          0               0   
58052         0       0              0          0  

We will drop the person with ID: 53298 (age 123) because it has zero on all values. So even if it is a real person, it just doesn't add any value to our segmentation.

In [13]:
# Drop row with id = 53298
data.drop(53298, inplace=True)